packages/peregrine/lib/talons/Accordion/useAccordion.js

```diff
@@ -0,0 +1,82 @@
+import {
+    Children,
+    cloneElement,
+    useCallback,
+    useEffect,
+    useState
+} from 'react';
+
+export const useAccordion = props => {
+    const { canOpenMultiple, children } = props;
+
+    const [openSectionIndicies, setOpenSectionIndicies] = useState(new Set([]));
+
+    const handleSectionClick = useCallback(
+        sectionIndex => {
+            const updatedIndicies = new Set(openSectionIndicies);
+
+            if (!openSectionIndicies.has(sectionIndex)) {
+                // The user wants to open this section.
+                if (!canOpenMultiple) {
+                    updatedIndicies.clear();
+                }
+
+                updatedIndicies.add(sectionIndex);
+            } else {
+                // The user wants to close this section.
+                updatedIndicies.delete(sectionIndex);
+            }
+
+            setOpenSectionIndicies(updatedIndicies);
+        },
+        [canOpenMultiple, openSectionIndicies, setOpenSectionIndicies]
+    );
```

Comment:
You're on the right track here. Unfortunately, this callback will be recreated on every state update, because it depends on `openSectionIndices`, which is the very same state it's updating.

To work around this, use the functional update signature of `setOpenSectionIndices`:

```es6
const handleSectionClick = useCallback(
  index => {
    setOpenSectionIndices(prevIndices => {
      const nextIndices = new Set(prevIndices)

      if (!prevIndices.has(index)) {
        if (!canOpenMultiple) {
          nextIndices.clear()
        }

        nextIndices.add(index)
      } else {
        nextIndices.delete(index)
      }

      return nextIndices
    })
  },
  [canOpenMultiple, setOpenSectionIndices]
)
```